platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileWriterPbj.java

```diff
@@ -0,0 +1,307 @@
+            throws IOException {
+        this.dataItemSerializer = dataItemSerializer;
+        this.path = createDataFilePath(filePrefix, dataFileDir, index, creationTime, extension);
+        this.lockFilePath = getLockFilePath(path);
+        if (Files.exists(lockFilePath)) {
+            throw new IOException("Tried to start writing to data file [" + path + "] when lock file already existed");
+        }
+        metadata = new DataFileMetadata(
+                0, // data item count will be updated later in finishWriting()
+                index,
+                creationTime,
+                dataItemSerializer.getCurrentDataVersion());
+        writingChannel = FileChannel.open(
+                path, StandardOpenOption.CREATE_NEW, StandardOpenOption.READ, StandardOpenOption.WRITE);
+        Files.createFile(lockFilePath);
+        writeHeader();
+    }
+
+    /**
+     * Maps the writing byte buffer to the given position in the file. Byte buffer size is always
+     * {@link #MMAP_BUF_SIZE}. Previous mapped byte buffer, if not null, is released.
+     *
+     * @param newMmapPos new mapped byte buffer position in the file, in bytes
+     * @throws IOException if I/O error(s) occurred
+     */
+    private void moveWritingBuffer(final long newMmapPos) throws IOException {
+        mmapPositionInFile = newMmapPos;
+        if (writingMmap != null) {
+            DataFileCommon.closeMmapBuffer(writingMmap);
+        }
+        writingMmap = writingChannel.map(MapMode.READ_WRITE, mmapPositionInFile, MMAP_BUF_SIZE);
+        writingPbjData = BufferedData.wrap(writingMmap);
+    }
+
+    // Future work: make it private
+    protected void writeHeader() throws IOException {
+        writingHeaderMmap = writingChannel.map(MapMode.READ_WRITE, 0, 1024);
+        writingHeaderPbjData = BufferedData.wrap(writingHeaderMmap);
+
+        metadata.writeTo(writingHeaderPbjData);
+
+        // prepare to write data items
+        moveWritingBuffer(writingHeaderPbjData.position());
+    }
+
+    public DataFileType getFileType() {
+        return DataFileType.PBJ;
+    }
+
+    /** Get the path for the file being written. Useful when needing to get a reader to the file. */
+    public Path getPath() {
+        return path;
+    }
+
+    /**
+     * Get file metadata for the written file.
+     *
+     * @return data file metadata
+     */
+    public DataFileMetadata getMetadata() {
+        return metadata;
+    }
+
+    /**
+     * Write a data item copied from another file like during merge. The data item serializer
+     * copyItem() method will be called to give it a chance to pass the data for or upgrade the
+     * serialization as needed.
+     *
+     * @param dataItemData a buffer containing the item's data
+     * @return New data location in this file where it was written
+     * @throws IOException If there was a problem writing the data item
+     */
+    public synchronized long writeCopiedDataItem(final Object dataItemData)
```

Comment:
`writeCopiedDataItem` and `storeDateItem` method are very similar. Have you considered extracting common parts of these methods into a separate method to reuse it?

Reply:
Yeah, they are very similar... I guess they can be refactored to a single method with a couple lambdas for real calls to copy/serialize data, but performance wise it will be slower. Existing `DataFileWriter` has the same issue. I'd leave it as is for now and check performance impact later.